File: Example/macOS-Objective-C/PaintingItem.m
//
//  PaintingItem.m
//  macOS-Objective-C
//
//  Created by Felix Herrmann on 23.10.21.
//

#import "PaintingItem.h"
@import UIImageColorsObjc;


@interface PaintingItem ()

@property NSView *_imageView;

@end


@implementation PaintingItem


#pragma mark - Setter

- (void)setImage:(NSImage *)image {
    [self._imageView.layer setContents:image];
    [image getColorsWithQuality:UIImageColorsScaleQualityHigh completion:^(NSImageColors * _Nullable colors) {
        if(colors != nil) {
            [self.view.layer setBackgroundColor:colors.background.CGColor];
            [self.mainLabel setTextColor:colors.primary];
            [self.secondaryLabel setTextColor:colors.secondary];
            [self.detailLabel setTextColor:colors.detail];
        } else {
            [self.view.layer setBackgroundColor:nil];
            [self.mainLabel setTextColor:NSColor.labelColor];
            [self.secondaryLabel setTextColor:NSColor.labelColor];
            [self.detailLabel setTextColor:NSColor.labelColor];
        }
    }];
}


#pragma mark - ViewController

- (void)loadView {
    self.view = [[NSView alloc] init];
    [self.view setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];
    [self.view setWantsLayer:true];
    
    self._imageView = [[NSView alloc] init];
    [self._imageView setLayer:[[CALayer alloc] init]];
    [self._imageView.layer setContentsGravity:kCAGravityResizeAspectFill];
    [self._imageView setTranslatesAutoresizingMaskIntoConstraints:false];
    
    self.mainLabel = [[NSLabel alloc] init];
    [self.mainLabel setFont:[NSFont systemFontOfSize:15 weight:NSFontWeightSemibold]];
    [self.mainLabel setTranslatesAutoresizingMaskIntoConstraints:false];
    
    self.secondaryLabel = [[NSLabel alloc] init];
    [self.secondaryLabel setFont:[NSFont systemFontOfSize:12 weight:NSFontWeightSemibold]];
    [self.secondaryLabel setTranslatesAutoresizingMaskIntoConstraints:false];
    
    self.detailLabel = [[NSLabel alloc] init];
    [self.detailLabel setFont:[NSFont systemFontOfSize:12 weight:NSFontWeightSemibold]];
    [self.detailLabel setTranslatesAutoresizingMaskIntoConstraints:false];
    
    NSLayoutGuide *layoutGuide = [[NSLayoutGuide alloc] init];
    [self.view addLayoutGuide:layoutGuide];
    
    [self.view addSubview:self._imageView];
    [self.view addSubview:self.mainLabel];
    [self.view addSubview:self.secondaryLabel];
    [self.view addSubview:self.detailLabel];
    
    [NSLayoutConstraint activateConstraints:@[
        [layoutGuide.leadingAnchor constraintEqualToAnchor:self._imageView.trailingAnchor constant:20],
        [layoutGuide.trailingAnchor constraintLessThanOrEqualToAnchor:self.view.trailingAnchor constant:-20],
        [layoutGuide.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor],
        
        [self._imageView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [self._imageView.topAnchor constraintEqualToAnchor:self.view.topAnchor],
        [self._imageView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
        [self._imageView.widthAnchor constraintEqualToAnchor:self._imageView.heightAnchor],
        
        [self.mainLabel.leadingAnchor constraintEqualToAnchor:layoutGuide.leadingAnchor],
        [self.mainLabel.topAnchor constraintEqualToAnchor:layoutGuide.topAnchor],
        [self.mainLabel.trailingAnchor constraintLessThanOrEqualToAnchor:layoutGuide.trailingAnchor],
        
        [self.secondaryLabel.leadingAnchor constraintEqualToAnchor: layoutGuide.leadingAnchor],
        [self.secondaryLabel.topAnchor constraintEqualToAnchor: self.mainLabel.bottomAnchor constant:4],
        [self.secondaryLabel.bottomAnchor constraintEqualToAnchor: layoutGuide.bottomAnchor],
        
        [self.detailLabel.leadingAnchor constraintEqualToAnchor: self.secondaryLabel.trailingAnchor constant: 4],
        [self.detailLabel.firstBaselineAnchor constraintEqualToAnchor: self.secondaryLabel.firstBaselineAnchor],
        [self.detailLabel.trailingAnchor constraintLessThanOrEqualToAnchor: layoutGuide.trailingAnchor]
    ]];
}

@end

File: Example/macOS/PaintingItem.swift
//
//  PaintingItem.swift
//  macOS
//
//  Created by Felix Herrmann on 22.10.21.
//

import AppKit
import UIImageColors

final class PaintingItem: NSCollectionViewItem {
    
    
    // MARK: - Properties
    
    private let _imageView = NSView()
    let mainLabel = NSLabel()
    let secondaryLabel = NSLabel()
    let detailLabel = NSLabel()
    
    var image: NSImage? {
        didSet {
            _imageView.layer?.contents = image
            image?.getColors { colors in
                if let colors = colors {
                    self.view.layer?.backgroundColor = colors.background.cgColor
                    self.mainLabel.textColor = colors.primary
                    self.secondaryLabel.textColor = colors.secondary
                    self.detailLabel.textColor = colors.detail
                } else {
                    self.view.layer?.backgroundColor = nil
                    self.mainLabel.textColor = .labelColor
                    self.secondaryLabel.textColor = .labelColor
                    self.detailLabel.textColor = .labelColor
                }
            }
        }
    }
    
    
    // MARK: - ViewController
    
    override func loadView() {
        view = NSView()
        view.autoresizingMask = [.width, .height]
        view.wantsLayer = true
        
        _imageView.layer = CALayer()
        _imageView.layer?.contentsGravity = .resizeAspectFill
        _imageView.translatesAutoresizingMaskIntoConstraints = false
        
        mainLabel.font = .systemFont(ofSize: 15, weight: .semibold)
        mainLabel.translatesAutoresizingMaskIntoConstraints = false
        
        secondaryLabel.font = .systemFont(ofSize: 12, weight: .semibold)
        secondaryLabel.translatesAutoresizingMaskIntoConstraints = false
        
        detailLabel.font = .systemFont(ofSize: 12, weight: .semibold)
        detailLabel.translatesAutoresizingMaskIntoConstraints = false
        
        let layoutGuide = NSLayoutGuide()
        view.addLayoutGuide(layoutGuide)
        
        view.addSubview(_imageView)
        view.addSubview(mainLabel)
        view.addSubview(secondaryLabel)
        view.addSubview(detailLabel)
        
        NSLayoutConstraint.activate([
            layoutGuide.leadingAnchor.constraint(equalTo: _imageView.trailingAnchor, constant: 20),
            layoutGuide.trailingAnchor.constraint(lessThanOrEqualTo: view.trailingAnchor, constant: -20),
            layoutGuide.centerYAnchor.constraint(equalTo: view.centerYAnchor),
            
            _imageView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            _imageView.topAnchor.constraint(equalTo: view.topAnchor),
            _imageView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
            _imageView.widthAnchor.constraint(equalTo: _imageView.heightAnchor),
            
            mainLabel.leadingAnchor.constraint(equalTo: layoutGuide.leadingAnchor),
            mainLabel.topAnchor.constraint(equalTo: layoutGuide.topAnchor),
            mainLabel.trailingAnchor.constraint(lessThanOrEqualTo: layoutGuide.trailingAnchor),
            
            secondaryLabel.leadingAnchor.constraint(equalTo: layoutGuide.leadingAnchor),
            secondaryLabel.topAnchor.constraint(equalTo: mainLabel.bottomAnchor, constant: 4),
            secondaryLabel.bottomAnchor.constraint(equalTo: layoutGuide.bottomAnchor),
            
            detailLabel.leadingAnchor.constraint(equalTo: secondaryLabel.trailingAnchor, constant: 4),
            detailLabel.firstBaselineAnchor.constraint(equalTo: secondaryLabel.firstBaselineAnchor),
            detailLabel.trailingAnchor.constraint(lessThanOrEqualTo: layoutGuide.trailingAnchor)
        ])
    }
}
